### ISSUE TITLE
Dotnet 9 Functions IHostApplicationBuilder

### ISSUE THREAD

**Issue by cliedeman** [getsentry/sentry-dotnet#3773]
### Problem Statement

Dotnet 9 has included support for using the IHostApplicationBuilder style for app creation

https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide?tabs=ihostapplicationbuilder%2Cwindows#start-up-and-configuration

With the new style we dont have access to the `HostBuilderContext`

```
var builder = FunctionsApplication.CreateBuilder(args);

builder.UseSentry(<missing>, options =>
{
    ...
});
```

Temporary workaround:

```
var builder = FunctionsApplication.CreateBuilder(args);

var context = new HostBuilderContext(new Dictionary<object, object>())
{
    Configuration = builder.Configuration,
};

builder.UseSentry(context, options =>
{
    ...
});
```

I see 2 obvious solutions:

Either add a builder extension with using `FunctionsApplicationBuilder`

or do a cast

```
    public static IHostApplicationBuilder UseSentry(
        this IFunctionsWorkerApplicationBuilder builder,
        Action<SentryAzureFunctionsOptions>? optionsConfiguration)
    {
        builder.UseMiddleware<SentryFunctionsWorkerMiddleware>();
        IServiceCollection services = builder.Services;
        IConfigurationSection section;
        if (builder is IHostApplicationBuilder applicationBuilder)
        {
            section = applicationBuilder.Configuration.GetSection("Sentry");
        }
        else
        {
            throw new ArgumentException("builder is not a IHostApplicationBuilder");
        }
        
        services.AddSingleton<IConfigureOptions<SentryAzureFunctionsOptions>>((Func<IServiceProvider, IConfigureOptions<SentryAzureFunctionsOptions>>) (_ => (IConfigureOptions<SentryAzureFunctionsOptions>) new SentryAzureFunctionsOptionsSetup((IConfiguration) section)));
        if (optionsConfiguration != null)
            services.Configure<SentryAzureFunctionsOptions>(optionsConfiguration);
        services.AddLogging();
        services.AddSingleton<ILoggerProvider, SentryAzureFunctionsLoggerProvider>();
        services.AddSingleton<IConfigureOptions<SentryAzureFunctionsOptions>, SentryAzureFunctionsOptionsSetup>();
        services.AddSentry<SentryAzureFunctionsOptions>();
        return builder;
    }
```

**Comment by bruno-garcia** [comment]
Thanks for raising this. We're working on improved .NET 9.0 support now on `main` branch as part of our 5.0.0 release.

Would you consider opening a PR?

**Comment by cliedeman** [comment]
Sure. which approach should I use?

**Comment by jamescrosswell** [comment]
It looks like it'd be easy to break with the cast option if someone called this incorrectly from an app that used the older `IHostBuilder`:
```csharp
        else
        {
            throw new ArgumentException("builder is not a IHostApplicationBuilder");
        }
```

Ideally we wouldn't find out at runtime what we could prevent with static types.

`SentryFunctionsWorkerApplicationBuilderExtensions` doesn't actually need `HostBuilderContext` or an `IHostApplicationBuilder`... ultimately the dependency it's after is `IConfiguration`. Maybe it makes sense to have one overload (where the real work happens) that takes an `IConfiguration` parameter... and then wrap this with some other overloads that are built to work specifically with either `HostBuilderContext` or `IHostApplicationBuilder`?

**Comment by GetoXs** [comment]
Hi guys,
How about change extension "base" interface from `IFunctionsWorkerApplicationBuilder` to `IHostApplicationBuilder`? It would simplify code, and it could be reusable in other sentry integrations (not only azure function).

```c#

	public static IHostApplicationBuilder UseSentry(
		this IHostApplicationBuilder builder,
		Action<SentryAzureFunctionsOptions>? optionsConfiguration)
	{

		var services = builder.Services;
		var section = builder.Configuration.GetSection("Sentry");
#if NET8_0_OR_GREATER
        services.AddSingleton<IConfigureOptions<SentryAzureFunctionsOptions>>(_ =>
            new SentryAzureFunctionsOptionsSetup(section)
        );
#else
        services.Configure<SentryAzureFunctionsOptions>(options =>
            section.Bind(options));
#endif
		...
	}
```

**Comment by jamescrosswell** [comment]
> How about change extension "base" interface from `IFunctionsWorkerApplicationBuilder` to `IHostApplicationBuilder`?

I don't think that would work with [the old way of initialising Azure Functions](https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide?tabs=hostbuilder,windows#start-up-and-configuration) though right?

We could provide extensions for both `IFunctionsWorkerApplicationBuilder` and `IHostApplicationBuilder`in the public API. Under the hood, it's probably worth refactoring an internal method that performs the same initialisation for both however (DRY), which takes an `IConfiguration` argument. 

I think the method itself will always be Azure Functions specific, since in the implementation it sets up Azure Functions specific middleware:
https://github.com/getsentry/sentry-dotnet/blob/bd62cf0421cc7a5d7b7804f8461fae00d47d43b8/src/Sentry.Azure.Functions.Worker/SentryFunctionsWorkerApplicationBuilderExtensions.cs#L37

**Comment by aritchie** [comment]
@jamescrosswell We could implement the UseMiddleware & further down from there, Use functionality on IHostApplicationBuilder, but I don't feel this is a good idea.  

https://github.com/Azure/azure-functions-dotnet-worker/blob/main/src/DotNetWorker.Core/Hosting/FunctionsWorkerApplicationBuilder.cs
